### ISSUE TITLE
Deadlock when renaming non empty folder

### ISSUE THREAD

**Issue by kalemas** [yandex-cloud/geesefs#53]
Hi, we experience blocking for geesefs mounted filesystem when renaming a folder (300 files, 125 GB). It was discussed on support thread at yandex cloud, ticket number 166384642671346, but as also related to geesefs itself we came here.

The behavior: after entering `mv match/Editing/MIX_Match_DEL match/Editing/MIX_Match_DELME` it hangs, `ls match/` succeed but `ls match/Editing` hangs. Another folders are renamed successfully, so it is only single folder that is affected with this issue been discovered (`storage.yandexcloud.net/match/Editing/MIX_Match_DEL`).

```
> geesefs --version
geesefs version 0.31.8
```

The stack trace that you may need, taken while filesystem is affected with this issue: [grs.txt](https://github.com/yandex-cloud/geesefs/files/9732542/grs.txt).

The log for geesefs itself contains a production data, that should not be published here. Is there chance that you would access to corresponding support thread (166384642671346) at yandex cloud?

**Comment by vitalif** [comment]
Ok, I checked your support thread. I see there is an exception in the log you posted there? `2022/09/29 16:20:12.689303 main.ERROR stacktrace from panic: double insert of` - this one. Does it still reproduce when you try to rename the directory?

**Comment by kalemas** [comment]
Yes, the exception as follows:

```
2022/10/10 18:04:10.924483 main.ERROR stacktrace from panic: double insert of Editing/MIX_Match_DELME/MIX/Output/***/final/***/***/..
goroutine 43195 [running]:
runtime/debug.Stack(0xc00073b658, 0xf7baa0, 0xc001323200)
        /opt/hostedtoolcache/go/1.16.15/x64/src/runtime/debug/stack.go:24 +0x9f
github.com/yandex-cloud/geesefs/api/common.LogPanic(0xc00073bf10)
        /home/runner/work/geesefs/geesefs/api/common/panic_logger.go:32 +0x76
panic(0xf7baa0, 0xc001323200)
        /opt/hostedtoolcache/go/1.16.15/x64/src/runtime/panic.go:965 +0x1b9
github.com/yandex-cloud/geesefs/internal.(*Inode).insertChildUnlocked(0xc00130ad80, 0xc00130b200)
        /home/runner/work/geesefs/geesefs/internal/dir.go:878 +0x605
github.com/yandex-cloud/geesefs/internal.(*Goofys).insertInode(0xc0003dca20, 0xc00130ad80, 0xc00130b200)
        /home/runner/work/geesefs/geesefs/internal/goofys.go:1043 +0x65
github.com/yandex-cloud/geesefs/internal.(*Inode).doMkDir(0xc00130ad80, 0x114319b, 0x2, 0xc00073b901)
        /home/runner/work/geesefs/geesefs/internal/dir.go:1100 +0x24f
github.com/yandex-cloud/geesefs/internal.renameRecursive(0xc00145aa80, 0xc00130ad80, 0x114319b, 0x2)
        /home/runner/work/geesefs/geesefs/internal/dir.go:1372 +0x46
github.com/yandex-cloud/geesefs/internal.renameRecursive(0xc00145a780, 0xc00130a900, 0xc000a74a49, 0x2a)
        /home/runner/work/geesefs/geesefs/internal/dir.go:1390 +0x1c5
github.com/yandex-cloud/geesefs/internal.renameRecursive(0xc00145a180, 0xc00130a480, 0xc002078276, 0x2)
        /home/runner/work/geesefs/geesefs/internal/dir.go:1390 +0x1c5
github.com/yandex-cloud/geesefs/internal.renameRecursive(0xc002c2db00, 0xc0012d3680, 0xc002078130, 0x5)
        /home/runner/work/geesefs/geesefs/internal/dir.go:1390 +0x1c5
github.com/yandex-cloud/geesefs/internal.renameRecursive(0xc000f92c00, 0xc0011a9200, 0xc00003a761, 0x1e)
        /home/runner/work/geesefs/geesefs/internal/dir.go:1390 +0x1c5
github.com/yandex-cloud/geesefs/internal.renameRecursive(0xc008dbcf00, 0xc0011a8d80, 0xc00bdbe94a, 0x6)
        /home/runner/work/geesefs/geesefs/internal/dir.go:1390 +0x1c5
github.com/yandex-cloud/geesefs/internal.renameRecursive(0xc008dbc600, 0xc0011a8900, 0xc0006b8d56, 0x3)
        /home/runner/work/geesefs/geesefs/internal/dir.go:1390 +0x1c5
github.com/yandex-cloud/geesefs/internal.renameRecursive(0xc000dfc780, 0xc0007b2000, 0xc0017b0cf0, 0xf)
        /home/runner/work/geesefs/geesefs/internal/dir.go:1390 +0x1c5
github.com/yandex-cloud/geesefs/internal.(*Inode).Rename(0xc0007b2000, 0xc0017b0ce0, 0xd, 0xc0007b2000, 0xc0017b0cf0, 0xf, 0x0, 0x0)
        /home/runner/work/geesefs/geesefs/internal/dir.go:1361 +0x4a5
github.com/yandex-cloud/geesefs/internal.(*Goofys).Rename(0xc0003dca20, 0x133ca78, 0xc0008b93b0, 0xc000d94040, 0x0, 0x0)        /home/runner/work/geesefs/geesefs/internal/goofys.go:1616 +0x205
github.com/yandex-cloud/geesefs/api/common.FusePanicLogger.Rename(0x134cd60, 0xc0003dca20, 0x133ca78, 0xc0008b93b0, 0xc000d94040, 0x0, 0x0)
        /home/runner/work/geesefs/geesefs/api/common/panic_logger.go:89 +0x8c
github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).handleOp(0xc0004bc960, 0xc00029b5f0, 0x133ca78, 0xc0008b93b0, 0xf4b5e0, 0xc000d94040)
        /home/runner/go/pkg/mod/github.com/vitalif/fusego@v0.0.0-20220702223751-e304003d79bf/fuseutil/file_system.go:187 +0x104
created by github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).ServeOps
        /home/runner/go/pkg/mod/github.com/vitalif/fusego@v0.0.0-20220702223751-e304003d79bf/fuseutil/file_system.go:123 +0x1a5

2022/10/10 18:04:10.924541 fuse.DEBUG Op 0x000008d4        connection.go:507] -> Error: "input/output error"
2022/10/10 18:04:10.924576 fuse.ERROR *fuseops.RenameOp error: input/output error
```

Its fresh, in sense of reproducibility.

**Comment by vitalif** [comment]
Hi. Fixed it in https://github.com/yandex-cloud/geesefs/releases/tag/v0.31.9

**Comment by kalemas** [comment]
unfortunately there is still deadlocks when working with geesefs, unfortunately panic logger does not commit operation and resource? timestamp, so it is much harden trace down this issue root cause:

```
2022/11/03 13:34:36.734800 main.ERROR Error reading 0 +57344 of folder/date/task/prop_status.png: NoSuchKey: The specified key does not exist.
        status code: 404, request id: 0ee40955f643c15f, host id:
2022/11/03 13:34:36.762420 main.WARNING File folder/date/task/prop_status.png is deleted or resized remotely, discarding local changes
2022/11/03 13:34:36.735071 main.ERROR Error reading 0 +57344 of folder/date/task/Character_prop1.png: NoSuchKey: The specified key does not exist.
        status code: 404, request id: 64da09ca3d6b6658, host id:
2022/11/03 13:34:36.763391 main.WARNING File folder/date/task/Character_prop1.png is deleted or resized remotely, discarding local changes
2022/11/03 13:34:36.749909 main.ERROR Error reading 0 +57344 of folder/date/task/prop1.png: NoSuchKey: The specified key does not exist.
        status code: 404, request id: db8b293a30f2e859, host id:
2022/11/03 13:34:36.763426 main.WARNING File folder/date/task/prop1.png is deleted or resized remotely, discarding local changes2022/11/07 07:21:28.139820 main.ERROR stacktrace from panic: deref inode 29396 (Assets/StudioLibrary/Trash/tmp (31).anim/sequence) by 3 from 2
goroutine 1256094 [running]:
runtime/debug.Stack(0xc000091bd8, 0xf7baa0, 0xc0048ce8f0)
        /opt/hostedtoolcache/go/1.16.15/x64/src/runtime/debug/stack.go:24 +0x9f
github.com/yandex-cloud/geesefs/api/common.LogPanic(0xc000091f10)
        /home/runner/work/geesefs/geesefs/api/common/panic_logger.go:32 +0x76
panic(0xf7baa0, 0xc0048ce8f0)
        /opt/hostedtoolcache/go/1.16.15/x64/src/runtime/panic.go:965 +0x1b9
github.com/yandex-cloud/geesefs/internal.(*Inode).DeRef(0xc003651500, 0x3, 0x72d4)
        /home/runner/work/geesefs/geesefs/internal/handles.go:361 +0x3a8
github.com/yandex-cloud/geesefs/internal.(*Goofys).ForgetInode(0xc00013c000, 0x133ca78, 0xc00ba7fb90, 0xc001eda930, 0x0, 0x18)
        /home/runner/work/geesefs/geesefs/internal/goofys.go:1080 +0xd4
github.com/yandex-cloud/geesefs/api/common.FusePanicLogger.ForgetInode(0x134cd60, 0xc00013c000, 0x133ca78, 0xc00ba7fb90, 0xc001eda930, 0x0, 0x0)
        /home/runner/work/geesefs/geesefs/api/common/panic_logger.go:61 +0x89
github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).handleOp(0xc000081980, 0xc000432680, 0x133ca78, 0xc00ba7fb90, 0xf4af20, 0xc000f5de00)        /home/runner/go/pkg/mod/github.com/vitalif/fusego@v0.0.0-20220702223751-e304003d79bf/fuseutil/file_system.go:160 +0xb58
created by github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).ServeOps
        /home/runner/go/pkg/mod/github.com/vitalif/fusego@v0.0.0-20220702223751-e304003d79bf/fuseutil/file_system.go:123 +0x1a5

2022/11/07 07:21:28.180459 fuse.ERROR *fuseops.BatchForgetOp error: input/output error
2022/11/07 07:54:54.265959 s3.WARNING Conflict detected: server-side ETag or size of Assets/Location/Asset/Textures/Asset_1007_Color.exr ("64e7e8a700d6e0316754445020e35ee2", 2566805) differs from local ("989f6ce8ac75a1c9147284db759db48d", 9837600). File is changed remotely, dropping cache
2022/11/07 08:27:58.512139 s3.WARNING Conflict detected: server-side ETag or size of Assets/Location/Asset/Textures/Asset_1007_Color.exr ("64e7e8a700d6e0316754445020e35ee2", 2566805) differs from local ("989f6ce8ac75a1c9147284db759db48d", 9837600). File is changed remotely, dropping cache
```

we will try to investigate further and start new issue in case we would identify it

**Comment by kalemas** [comment]
it is second case after we updated on 0.31.9

**Comment by vitalif** [comment]
"panic: deref inode" was recently fixed in #55, in 0.32.1 :)
